### PR TITLE
fix: match operation without topic only when it was delivered by device control topic

### PR DESCRIPTION
**Pull request by Ruadhri17**
## Proposed changes
This PR fixes issue introduced with #3198. Previously we always matched operation that had no topic provided in the operation file as we assumed the default one is device control topic. Now operation can be run with any topic that user provides in the operation file so this rule no longer applies. Operation without provided topic is added to operation handlers no matter which topic delivered the message, wheras it should be added to handlers only if the message was delivered by device control topic.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 1]
```diff
@@ -57,11 +59,17 @@ impl Operations {
...
         let mut vec: Vec<(String, Operation)> = Vec::new();
         for op in self.operations.iter() {
             match (op.topic(), op.on_fragment()) {
-                (None, Some(on_fragment)) => vec.push((on_fragment, op.clone())),
+                (None, Some(on_fragment)) if C8yDeviceControlTopic::name(prefix) == topic_name => {
+                    vec.push((on_fragment, op.clone()))
+                }
                 (Some(topic), Some(on_fragment)) if topic == topic_name => {
                     vec.push((on_fragment, op.clone()))
                 }
```
Comment (didier-wenzek) [inline crates/core/c8y_api/src/smartrest/operations.rs:75]: An alternative way to fix the issue is to use `C8yDeviceControlTopic::name(prefix)` as a default value for `OnMessageExec::topic`. Doing so would be less error prone.

**Review by didier-wenzek — COMMENTED** [review]
As this fix an issue introduced by another PR, it would be good to add a test reproducing the case.

**Review by rina23q — COMMENTED** [review]
The fix looks good, however, unit test is missing.

File: crates/core/c8y_api/src/smartrest/operations.rs (hunk 1 above)
Comment (Ruadhri17) [reply crates/core/c8y_api/src/smartrest/operations.rs:75]: In therory I agree but I cannot use it as serde default and changing `.topic()` function that unwraps to `C8yDeviceControlTopic::name(prefix)` when topic is `None` requires to provide that prefix almost in every single function in `operations.rs` (also in places that does not need that information) so I find it inconvenient.

**Review by didier-wenzek — APPROVED** [review]
Approved

File: crates/core/c8y_api/src/smartrest/operations.rs [hunk 2]
```diff
@@ -393,7 +401,7 @@ pub enum InvalidCustomOperationHandler {
 
 #[cfg(test)]
 mod tests {
-    use std::io::Write;
+    use std::{io::Write, str::FromStr};
```
Comment (rina23q) [inline crates/core/c8y_api/src/smartrest/operations.rs:404]: Format it :)

**Review by rina23q — APPROVED** [review]
Looks good now. Just fix the format check, then it will be good to merge.

**Pull request merged**